include/swift/Reflection/ReflectionContext.h

```diff
@@ -378,30 +379,42 @@ class ReflectionContext
     return readPECOFFSections(ImageStart);
   }
 
-  template <typename T> bool readELFSections(RemoteAddress ImageStart) {
+  template <typename T>
+  bool readELFSections(RemoteAddress ImageStart,
+                       llvm::sys::MemoryBlock FileBuffer) {
+    if (!FileBuffer.base() || !FileBuffer.allocatedSize())
+      return false;
+
     auto Buf =
         this->getReader().readBytes(ImageStart, sizeof(typename T::Header));
 
     auto Hdr = reinterpret_cast<const typename T::Header *>(Buf.get());
     assert(Hdr->getFileClass() == T::ELFClass && "invalid ELF file class");
 
-    // From the header, grab informations about the section header table.
-    auto SectionHdrAddress = ImageStart.getAddressData() + Hdr->e_shoff;
-    auto SectionHdrNumEntries = Hdr->e_shnum;
-    auto SectionEntrySize = Hdr->e_shentsize;
+    uint64_t FileBufferStartAddress = (uint64_t)FileBuffer.base();
+    // From the header, grab information about the section header table.
+    uint64_t SectionHdrAddress = FileBufferStartAddress + Hdr->e_shoff;
+    uint16_t SectionHdrNumEntries = Hdr->e_shnum;
+    uint16_t SectionEntrySize = Hdr->e_shentsize;
+
+    if (SectionHdrNumEntries == 0)
+      return false;
+    if (FileBufferStartAddress + FileBuffer.allocatedSize() <
+        SectionHdrAddress + (SectionHdrNumEntries - 1) * SectionEntrySize +
+            SectionEntrySize)
+      return false;
 
     // Collect all the section headers, we need them to look up the
     // reflection sections (by name) and the string table.
-    std::vector<const typename T::Section *> SecHdrVec;
+    // We read the section headers from the FileBuffer, since they are
+    // not mapped in the child process.
+    std::vector<std::unique_ptr<const typename T::Section>> SecHdrVec;
     for (unsigned I = 0; I < SectionHdrNumEntries; ++I) {
-      auto SecBuf = this->getReader().readBytes(
-          RemoteAddress(SectionHdrAddress + (I * SectionEntrySize)),
-          SectionEntrySize);
-      if (!SecBuf)
-        return false;
-      auto SecHdr =
-          reinterpret_cast<const typename T::Section *>(SecBuf.get());
-      SecHdrVec.push_back(SecHdr);
+      auto SecHdr = std::make_unique<const typename T::Section>();
+      memcpy((void *)SecHdr.get(),
+             (void *)(SectionHdrAddress + (I * SectionEntrySize)),
+             SectionEntrySize);
```

Comment:
I think I asked this before, but I must have misplaced the reply: Why do we need to memcpy at all and can't just access the buffer directly?

Reply:
I did so for the string table but I thought that it wouldn't be safe to do for the entries of the section header table because we place them in a vector for lookup later, but looking at it again I see that it's safe to do the same here.